### DISCUSSION TITLE
Is it possible to use HERBS to visualize the fiber locations and probes of multiple subjects in the same figure?

### DISCUSSION THREAD

**Discussion by skrigel** [JingyiGF/HERBS#57] [Q&A]
We want to visualize all images from all subjects in the same figure, so we have to plot them all together in the same coordinate space. The idea is to match the position of each fiber to one atlas section, like in the images from the paper linked below (but in the atlas space rather than plotting by hand). We would also like to visualize each probe against the atlas.

Here's a link to the paper to get a better idea of what I'm talking about: https://pubmed.ncbi.nlm.nih.gov/37495689/

<img width="1292" alt="Screenshot 2024-01-30 at 10 19 04 AM" src="https://github.com/JingyiGF/HERBS/assets/140971083/56188422-4110-4018-b0ef-bc63546c64fd">

**Comment by JingyiGF** [comment]
Oooops, you ask something big without providing clear and detailed information. As you might know from other issues and discussions, my background is not in Neuroscience. So may I ask what do you mean by "subjects"? "fiber"? I am sorry I have no idea what this figure is about. Honestly, I do not have time to read the paper, and I have no access to read the full paper.  So I do not know what I can do to help you with the current info you provided. If you could explain a bit more what you are trying to do, with text and drawings, things will go easier I hope.

**Reply by matiasandina** [reply]
Hi, thanks for the quick response. I thought I might chime in to help clarify the question.

It's common in our field to run multiple subjects (i.e., mice) within the same experiment. Due to variability during surgeries, the implants (i.e., optic fibers, cannulas, probes) will never be in the exact same spot. 

It is common to report the placement of the implant for each animal. That is what the image from the paper is showing (no need to read the paper, it's just a basic representation of where probes landed for each subject in this particular experiment). As you can see in the image, there is a sort of cartoon of the brain drawing with AP levels in mm from Bregma, and each animal has a rectangle indicating the position of the tip of a cannula.

Ideally, we would quantify multiple animals using HERBS, save the coordinates, and plot coordinates from all animals according to the atlas plate they better correspond to (e.g, -1.5 mm from Bregma, or the equivalent in Allen space). 
I hope this further clarifies and I'm happy to provide more detail if needed!

Below is another cartoon example of what we are aiming for. Ideally, we would be able to do this using the output from HERBS (or a combination of the HERBS output with some scaling into another coordinate space). 

![image](https://github.com/JingyiGF/HERBS/assets/7494967/9afc0aa5-a784-40ff-96c4-0f85f6a4ce58)


or maybe even something like this

![image](https://github.com/JingyiGF/HERBS/assets/7494967/4923af57-c523-4976-ad76-92978bdb3a23)

Of note, these cartoons were generated programatically but, as you can probably guess from the magnitude of the coords, it doesn't actually match with the ones given by . We are happy to work with you in figuring out the equivalence between the two coords for the different renders of the allen atlas, since it should be a simple scaling, but we were thinking that an easier route might be to use whatever herbs already has.

What we are thinking is:

1. Quantify using HERBS
2. Export the probe into `.pkl` (or whatever other format)
3. Repeat 1 and 2 for all the animals
4. Load each animal export in a separate python session and extract the data
5. Visualize by maybe (?):
    a. Using atlas data from HERBS itself that is accessible from the API. This would also be amazing to do in 3D, kind of what you show in the cookbook for the 4 probes in the same animal, only it would be for different animals.
    b. Another strategy (e.g., export and scale outside HERBS)

We could do something along the lines of:

```
import pickle

# repeat for each animal and store into a data structure for proper plotting/export
with open(‘probe_export.pkl', 'rb') as file:
	my_probe= pickle.load(file)
# coords are given ML, AP, DV
# Which translate to x, AP, y in coronal space
insertion_vox = my_probe[‘data’]['insertion_vox']
terminus_vox = my_probe['data']['terminus_vox']

# somehow do the plotting here ? from herbs import the corresponding AP level plate to draw?
# or export coords before/after doing proper transformation 

```

We believe the information to make these plots to be available already in herbs, since clicking the probe report informs of all the regions where the probe goes through. It's not clear how to access these methods though (possibly due to lack of documentation, but maybe because methods are not exposed to the user). We are happy to make a tutorial for the use case if this is something that would be in the realm of possible / desireable.

**Reply by JingyiGF** [reply]
Ok, I see what you mean now. To be crystal clear, you would like to have all saved probes (from your work with different animals) to be visualized in the same 2d atlas slice if all probes are registered to the same atlas slice and also in 3d atlas, right? If this is the case, all probes can automatically be shown in 3d atlas. 

![image](https://github.com/JingyiGF/HERBS/assets/64961126/070394a3-c0c5-4f55-a6bc-4ed7787ff981)

To visualise in 2D atlas slice, I need to do a bit work on it, as shown in the screenshot as follows,
![image](https://github.com/JingyiGF/HERBS/assets/64961126/93635330-0247-4419-ad81-356b27c9a11d)

Hmmmm, the design was after users loaded all probes that were saved, users should link the probe object that want to compare and then click the 2D-button to visualise them in 2D atlas slice. But as you can see, it was under development ;-)

I think it was in progress some time ago, and I was distracted with other stuffs, so did not finish this functionality.

**Reply by JingyiGF** [reply]
The above is for register probes in volume atlas, if you want to register to PAXINOS, for example, you have to read https://github.com/JingyiGF/HERBS/blob/main/Tutorial/Atlas_Related/upload_slice_atlas.md. Not so many people tested this functionality for it, so if you found something unusual, I will fix :-)

**Reply by matiasandina** [reply]
> Hmmmm, the design was after users loaded all probes that were saved, users should link the probe object that want to compare and then click the 2D-button to visualise them in 2D atlas slice. But as you can see, it was under development ;-)

Yes, this would be it, but ideally we would do it as a script or set of function calls rather than clicking on each animal file location, loading probes, linking them, etc. I'm not exactly sure how to solve the "probe in plate problem" since the probe usually spans more than one section, but showing all sections with annotated probe would be good enough.

I'm not sure how much of a development from your end this would be. We are happy to help if given some pointers for the way you would approach this

**Reply by JingyiGF** [reply]
Ah, now I see what you mean. So you are saying all probes are not lying perfect on the same cut, but you would like to visualize them on to a perfect cut. Right? This is a totally different story than what I thought. In this case, you have to do it with code. HERBS 2D will not support this. What you probably want to do,
1. save the voxel coordinates (x, y, z) for the insertion and terminus.
2. plot the atlas cut, for example, for coronal cut, plot a line go through insertion and terminus's (x, z).
This is just for visualisation, as lenght traveled in each brain region of a probe will definitely be different when plotting the probe that cross multiple sections into a single section.

**Reply by matiasandina** [reply]
Not exactly. What I am saying is this: The thickness of the probe is greater than the section (400 um vs 100 um). In addition, there will never be perfect alignment between the insertion angle and the cut.

That means we want to trace the probe in whatever sections it appears (multiple sections) and visualize the probe in all those sections using a programmatic approach instead of the GUI. Each animal would have a set of atlas plates with portions of probe in 2D. Because animals are targeted to the same location, we would show multiple animals for each AP as shown in the plots above.

It's somewhat clear how to use the x,z cords. But it's not clear how to find the atlas cut for the specific location of the probe in herbs. How do you find and plot that atlas level ?

Is it possible to plot the atlas reference plate, as shown in the reference atlas instead of the tissue pics, with delineated regions in schematic fashion?

**Reply by JingyiGF** [reply]
Hmmm, ok. If you go to the folder where you save your atlas, you will see there are multiple files for the processed atlas. Find the pre-xxx pickle files, load it. Atlas data are saved inside as dictionaries of numpy arrays, which in another word, a 3d matrix with numbers of color levers. The length of x direction is number of sagital cuts in the processed atlas. And length of y direction is the number of coronal cut and so on. Say A is the matrix of atlas data, you want data for coronal cut 35, then A[:, 35,:] is the data for the atlas slice. Is this the problem? Or? Coordinates related to bregma will not work here, the easiest is to use voxel coordinates. Fetching atlas slices is easy, unless you want to get a rotated cut like what you can do in HERBS. For rotated cuts, I have to make a API. 

Note: rotated cuts are numerical calculated that are not perpendicular to any axis. Whereas a perfect cut is perpendicular to a given axis.
